File: scripts/interfacing_C/rotacion.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h> 


    typedef struct {
      float m[3][3];
    } m3x3;

    typedef struct {
      float a[3];
    } v3;

  // !> matrix_rotation
  // !! Crea la matriz de rotación correspondiente a los tres ángulos de Euler
  // !! 
  // !! @param angles 
  // !! @return R

    m3x3 matrix_rotation(float angles[3]){

        m3x3 R;

        float cx = cos(angles[0]); 
        float cy = cos(angles[1]); 
        float cz = cos(angles[2]);
        float sx = sin(angles[0]); 
        float sy = sin(angles[1]); 
        float sz = sin(angles[2]);


        R.m[0][0] = cx*cz - sx*cy*sz;
        R.m[0][1] = cx*sz + sx*cy*cz;
        R.m[0][2] = sx*sy;

        R.m[1][0] = -sx*cz - cx*cy*sz;
        R.m[1][1] = -sx*sz + cx*cy*cz;
        R.m[1][2] = cx*sy;

        R.m[2][0] = sy*sz;
        R.m[2][1] = -sy*cz;
        R.m[2][2] = cy;

        return R;
  }

  v3 matmul3(m3x3 A, float b[3]){

      static v3 a;

      for(int i=0; i < 3; i++){
            a.a[0] = A.m[0][0] * b[0] + A.m[0][1] * b[1] + A.m[0][2] * b[2];
            a.a[1] = A.m[1][0] * b[0] + A.m[1][1] * b[1] + A.m[1][2] * b[2];
            a.a[2] = A.m[2][0] * b[0] + A.m[2][1] * b[1] + A.m[2][2] * b[2];
          }
    //  printf("%6.3f %6.3f %6.3f \n",a[0],a[1],a[2]);

      return a;
  }



    float * rotate(float angles[3], float *v, int N){
        
        m3x3 R = matrix_rotation(angles);
        

        float* y = (float*)malloc(3*N*sizeof(float));
        v3 p;

        printf("%p\n",y);
        for(int i=0; i<N; i++){
            // p = &y[i*3];
            p = matmul3(R,&v[i*3]);
            y[i*3+0] = p.a[0];
            y[i*3+1] = p.a[1];
            y[i*3+2] = p.a[2];
            // printf("%6.3f %6.3f %6.3f \n",y[i*3+0],y[i*3+1],y[i*3+2]);
        }
        return y;


  }


int main(){

  float angle[3] = { 0.0f, 0.0f, 3.141592f/2.0f};

  m3x3 R;

  int NDIM = 2;
  float* v = (float*)malloc(3*NDIM*sizeof(float));

  float *y;

  for (size_t j = 0; j < NDIM; j++)
      {
            v[j*3+0] = (j*3+0)*1.0f;
            v[j*3+1] = (j*3+1)*1.0f;
            v[j*3+2] = (j*3+2)*1.0f;
      }

  v[0] = (1)*1.0f;
  v[1] = (0)*1.0f;
  v[2] = (0)*1.0f;
  v[3] = (0)*1.0f;
  v[4] = (1)*1.0f;  
  v[5] = (0)*1.0f;



  R = matrix_rotation(angle);

    for(int i = 0; i < 3; i++){
        for(int j = 0; j < 3; j++){
            printf("%6.3f",R.m[i][j]);
            printf("   ");
        }
        printf("\n");
    }

  y = rotate(angle,v,NDIM);
  printf("%6.3f %6.3f %6.3f \n",y[0],y[1],y[2]);
  printf("%6.3f %6.3f %6.3f \n",y[3],y[4],y[5]);
  // for (size_t i = 0; i < NDIM; i++)
  // {
  //    printf("%6.3f %6.3f %6.3f \n",y[i*3],y[i*3+1],y[i*3+2]);
  // }
  

}  

File: scripts/interfacing_F/rotacion.f90
module rotaciones

contains
  !> matrix_rotation
  !! Crea la matriz de rotación correspondiente a los tres ángulos de Euler
  !! 
  !! @param angles 
  !! @return R
  function matrix_rotation(angles) result(R)
    implicit none
    real(8), dimension(3), intent(IN) :: angles
    real(8), dimension(3,3) :: R
    real(8) :: cx, cy, cz, sx, sy, sz

    cx = cos(angles(1)); cy = cos(angles(2)); cz = cos(angles(3))
    sx = sin(angles(1)); sy = sin(angles(2)); sz = sin(angles(3))

    R(1,1) = cx*cz - sx*cy*sz
    R(1,2) = cx*sz + sx*cy*cz
    R(1,3) = sx*sy

    R(2,1) = -sx*cz - cx*cy*sz
    R(2,2) = -sx*sz + cx*cy*cz
    R(2,3) = cx*sy

    R(3,1) = sy*sz
    R(3,2) = -sy*cz
    R(3,3) = cy
  end function matrix_rotation

  !> rotate
  !!
  !! @param angles 
  !! @param v 
  !! @param N longitud del vector v 
  !! @return y
  function rotate(angles, v, N) result(y)
    implicit none
    integer, intent(IN) :: N
    real(8), dimension(3), intent(IN) :: angles
    real(8), dimension(3,N), intent(IN) :: v
    real(8), dimension(3,N) :: y

    y = matmul(matrix_rotation(angles), v)
  end function rotate

end module rotaciones

program test_rotation
  USE rotaciones !, only: rotate
  implicit none
  integer, parameter :: Ndim=1000
  real(8), dimension(3) :: angle
  real(8), dimension(3, Ndim) :: v, y

  integer :: n, Nloops,i,j
  real(8) :: time_begin, time_end

  real(8), dimension(3, 3) :: R


  ! Selecciono Ndim vectores tridimensionales al azar
  call RANDOM_NUMBER(v)
  Nloops = 10000

  CALL CPU_TIME ( time_begin )
  do n=1,Nloops
    call RANDOM_NUMBER(angle)
    y= rotate(angle, v, Ndim)
  end do
  CALL CPU_TIME ( time_end )

  write (*,'(A,1PE12.2,A)') '# Tiempo usado:', (time_end - time_begin)/(1.e-3*Nloops), ' milisegundos por 1000 rotaciones'

  ! Test a simple case

  ! angle(1) = 1.0
  ! angle(2) = 2.0
  ! angle(3) = 3.0

  ! R = matrix_rotation(angle)
  ! ! do i=1,3
  ! !  write(*,'(3F8.3)') R(i,1),R(i,2),R(i,3)    
  ! ! end do

  ! do j=0,4
  !     v(1,j+1) = j*3+0
  !     v(2,j+1) = j*3+1
  !     v(3,j+1) = j*3+2
  ! end do
  ! y= rotate(angle, v, 5)

  ! do i=1,5
  !   write(*,'(3F8.3)') y(1,i),y(2,i),y(3,i)    
  ! end do

end program test_rotation
